File: optimization-string-processing-problems-complete/src/algorithms/longestPalindromicSubstring.ts
```typescript
/**
 * src/algorithms/longestPalindromicSubstring.ts
 *
 * Problem: Longest Palindromic Substring
 * Given a string `s`, return the longest palindromic substring in `s`.
 *
 * A substring is a contiguous non-empty sequence of characters within a string.
 * A palindrome is a string that reads the same forwards and backward.
 *
 * Example 1:
 * Input: s = "babad"
 * Output: "bab" ("aba" is also a valid answer)
 *
 * Example 2:
 * Input: s = "cbbd"
 * Output: "bb"
 */

/**
 * Approach 1: Expand Around Center
 *
 * This approach iterates through each character of the string, treating it as a potential center
 * of a palindrome. Palindromes can be of two types:
 * 1. Odd length: centered around a single character (e.g., "aba" centered at 'b').
 * 2. Even length: centered between two characters (e.g., "abba" centered between the two 'b's).
 *
 * For each character, we expand outwards to find the longest palindrome. We do this twice:
 * once assuming an odd length palindrome (center `i`) and once assuming an even length palindrome
 * (center `i` and `i+1`).
 *
 * We keep track of the start index and length of the longest palindrome found so far.
 *
 * Time Complexity: O(N^2)
 *   - There are 2N-1 possible centers (N for odd length, N-1 for even length).
 *   - In the worst case, expanding around a center can take O(N) time (e.g., "aaaaa").
 *   - So, total time complexity is O(N * N) = O(N^2).
 *
 * Space Complexity: O(1)
 *   - We only use a few variables to store the current longest palindrome's bounds.
 */
export function longestPalindromeExpandAroundCenter(s: string): string {
  if (!s || s.length < 1) {
    return "";
  }

  let start = 0; // Start index of the longest palindrome found
  let maxLength = 0; // Length of the longest palindrome found

  /**
   * Helper function to expand around a center and find the length of the palindrome.
   * @param left The left pointer of the potential palindrome center.
   * @param right The right pointer of the potential palindrome center.
   * @returns The length of the palindrome found.
   */
  const expandAroundCenter = (l: number, r: number): number => {
    while (l >= 0 && r < s.length && s[l] === s[r]) {
      l--;
      r++;
    }
    // After the loop, l and r are one step beyond the actual palindrome boundaries.
    // The length is (r - 1) - (l + 1) + 1 = r - l - 1.
    return r - l - 1;
  };

  for (let i = 0; i < s.length; i++) {
    // Case 1: Odd length palindrome (centered at s[i])
    const len1 = expandAroundCenter(i, i);

    // Case 2: Even length palindrome (centered between s[i] and s[i+1])
    const len2 = expandAroundCenter(i, i + 1);

    // Get the maximum length between the two cases
    const currentMaxLen = Math.max(len1, len2);

    // If the current palindrome is longer than the longest found so far, update.
    if (currentMaxLen > maxLength) {
      maxLength = currentMaxLen;
      // Calculate the new start index.
      // For a palindrome of length `currentMaxLen` centered at `i` (or `i` and `i+1`):
      // Start = i - (currentMaxLen - 1) / 2
      // This formula works for both odd and even lengths due to integer division.
      // E.g., for "babad" (len 5, center at 'a' (index 2)): 2 - (5-1)/2 = 2 - 2 = 0. "babad"
      // E.g., for "baba" (len 4, center between 'a's (index 1 & 2)): 1 - (4-1)/2 = 1 - 1 = 0. "baba"
      // More precisely, for 'len', start = i - floor((len-1)/2) if center is i.
      // If the center is i,i+1, start = i - (len/2 -1).
      // The formula `i - (currentMaxLen - 1) / 2` works well to pinpoint the start index.
      start = i - Math.floor((currentMaxLen - 1) / 2);
    }
  }

  // Return the substring using the calculated start and maxLength.
  return s.substring(start, start + maxLength);
}

/**
 * Approach 2: Dynamic Programming
 *
 * Let `dp[i][j]` be a boolean value indicating whether the substring `s[i...j]` is a palindrome.
 *
 * Base Cases:
 * 1. `dp[i][i] = true` (Single characters are palindromes).
 * 2. `dp[i][i+1] = (s[i] === s[i+1])` (Two characters are a palindrome if they are equal).
 *
 * Recurrence Relation:
 * `dp[i][j] = (s[i] === s[j] && dp[i+1][j-1])`
 * A substring `s[i...j]` is a palindrome if its outer characters `s[i]` and `s[j]` are equal
 * AND the inner substring `s[i+1...j-1]` is also a palindrome.
 *
 * The `dp` table is filled from smaller substrings to larger ones. This means we iterate
 * by `length` of substring (`L`), from `L=1` to `N`. For each `L`, we iterate `i` from `0` to `N-L`.
 * Then `j = i + L - 1`.
 *
 * Time Complexity: O(N^2)
 *   - We fill an N x N DP table.
 *   - Each cell takes O(1) time to compute.
 *   - Total time: O(N^2).
 *
 * Space Complexity: O(N^2)
 *   - We use an N x N 2D array to store the DP states.
 *   - Total space: O(N^2).
 */
export function longestPalindromeDP(s: string): string {
  if (!s || s.length < 1) {
    return "";
  }

  const n = s.length;
  // dp[i][j] will be true if substring s[i...j] is a palindrome
  const dp: boolean[][] = Array(n).fill(0).map(() => Array(n).fill(false));

  let start = 0;
  let maxLength = 1; // Single characters are palindromes, so min length is 1

  // All substrings of length 1 are palindromes
  for (let i = 0; i < n; i++) {
    dp[i][i] = true;
  }

  // Check for substrings of length 2
  for (let i = 0; i < n - 1; i++) {
    if (s[i] === s[i + 1]) {
      dp[i][i + 1] = true;
      start = i;
      maxLength = 2;
    }
  }

  // Check for substrings of length L >= 3
  // L is the current length of the substring
  for (let L = 3; L <= n; L++) {
    // i is the starting index of the substring
    for (let i = 0; i <= n - L; i++) {
      // j is the ending index of the substring
      const j = i + L - 1;

      // s[i...j] is a palindrome if s[i] == s[j] AND s[i+1...j-1] is a palindrome
      if (s[i] === s[j] && dp[i + 1][j - 1]) {
        dp[i][j] = true;
        // If we found a longer palindrome, update start and maxLength
        if (L > maxLength) {
          start = i;
          maxLength = L;
        }
      }
    }
  }

  return s.substring(start, start + maxLength);
}
```

File: algorithm-string-processing-implementation-advanced/src/main/java/com/example/stringmanipulation/problems/LongestPalindromicSubstring.java
```java
package com.example.stringmanipulation.problems;

/**
 * Problem: Longest Palindromic Substring
 * Given a string s, find the longest palindromic substring in s.
 * You may assume that the maximum length of s is 1000.
 */
public class LongestPalindromicSubstring {

    /**
     * Approach 1: Brute Force
     * Generate all possible substrings and for each substring, check if it's a palindrome.
     * Keep track of the longest palindrome found.
     *
     * Time Complexity: O(N^3)
     *   - O(N^2) for generating all substrings (N*(N+1)/2 substrings).
     *   - O(N) for checking if each substring is a palindrome.
     * Space Complexity: O(1) (if not counting the output string storage, which would be O(N))
     *
     * @param s The input string.
     * @return The longest palindromic substring.
     */
    public String longestPalindromeBruteForce(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }
        String longest = "";
        for (int i = 0; i < s.length(); i++) {
            for (int j = i; j < s.length(); j++) {
                String sub = s.substring(i, j + 1);
                if (StringManipulationUtils.isPalindrome(sub) && sub.length() > longest.length()) {
                    longest = sub;
                }
            }
        }
        return longest;
    }

    /**
     * Approach 2: Expand Around Center
     * A palindrome is symmetric around its center. The center can be a single character (for odd-length palindromes like "aba")
     * or two characters (for even-length palindromes like "abba").
     * We iterate through each possible center (N for single char, N-1 for two chars) and expand outwards to find the longest palindrome.
     *
     * Time Complexity: O(N^2)
     *   - There are 2N-1 possible centers (N single characters, N-1 pairs of characters).
     *   - For each center, expanding outwards takes O(N) time in the worst case.
     * Space Complexity: O(1) (if not counting the output string storage, which would be O(N))
     *
     * @param s The input string.
     * @return The longest palindromic substring.
     */
    public String longestPalindromeExpandAroundCenter(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int start = 0; // Start index of the longest palindrome found
        int end = 0;   // End index of the longest palindrome found

        for (int i = 0; i < s.length(); i++) {
            // Case 1: Odd length palindrome (center is s[i])
            int len1 = expandAroundCenter(s, i, i);
            // Case 2: Even length palindrome (center is s[i] and s[i+1])
            int len2 = expandAroundCenter(s, i, i + 1);

            int len = Math.max(len1, len2);

            if (len > end - start + 1) { // If current palindrome is longer
                // Update start and end indices
                // For a length 'len', the new start index will be i - (len - 1) / 2
                // The new end index will be i + len / 2
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        // Return the substring using the found start and end indices
        return s.substring(start, end + 1);
    }

    /**
     * Helper method to expand around a given center (or pair of centers) and return the length
     * of the palindrome found.
     *
     * @param s The input string.
     * @param left The left pointer of the center.
     * @param right The right pointer of the center.
     * @return The length of the palindrome centered at (left, right).
     */
    private int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        // When the loop exits, left and right are one step outside the palindrome boundaries.
        // The length is (right - 1) - (left + 1) + 1 = right - left - 1.
        return right - left - 1;
    }


    /**
     * Approach 3: Dynamic Programming
     * Define `dp[i][j]` as a boolean indicating whether `s[i...j]` is a palindrome.
     *
     * Recurrence relation:
     * `dp[i][j] = (s.charAt(i) == s.charAt(j)) && dp[i+1][j-1]`
     *
     * Base cases:
     * 1. Single character: `dp[i][i] = true` (all single characters are palindromes).
     * 2. Two characters: `dp[i][i+1] = (s.charAt(i) == s.charAt(i+1))` (if characters are same, it's a palindrome).
     *
     * We fill the DP table by increasing substring length (or "gap" between i and j) to ensure `dp[i+1][j-1]`
     * is already computed.
     *
     * Time Complexity: O(N^2)
     *   - Filling an N x N DP table.
     * Space Complexity: O(N^2)
     *   - For the N x N boolean DP table.
     *
     * @param s The input string.
     * @return The longest palindromic substring.
     */
    public String longestPalindromeDP(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int n = s.length();
        boolean[][] dp = new boolean[n][n]; // dp[i][j] is true if s[i...j] is a palindrome

        int start = 0; // Start index of the longest palindrome found
        int maxLength = 1; // Length of the longest palindrome found (minimum is 1 for single char)

        // All single characters are palindromes
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        // Check for palindromes of length 2
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLength = 2;
            }
        }

        // Check for palindromes of length 3 or more
        // 'len' is the current length of the substring we are checking
        for (int len = 3; len <= n; len++) {
            // 'i' is the starting index
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1; // 'j' is the ending index

                // s[i...j] is a palindrome if s[i] == s[j] AND s[i+1...j-1] is a palindrome
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    if (len > maxLength) {
                        start = i;
                        maxLength = len;
                    }
                }
            }
        }

        return s.substring(start, start + maxLength);
    }
}
```